### DISCUSSION TITLE
error: cannot create filelock for package(git)!

### DISCUSSION THREAD

**Discussion by Freed-Wu** [xmake-io/xmake#5699] [Q&A] [answered]
In some case

```
error: cannot create filelock for package(git)!
```

What does it mean?

**Comment by waruqi** [comment]
please provide whole logs and a reproducible process.

**Reply by Freed-Wu** [reply]
I try to run xmake in `nix-build -A tmux-rime`:

```nix
{ lib
, stdenv
, xmake
, fetchFromGitHub
, unzip
, glib
, librime
, pkg-config
}:
stdenv.mkDerivation rec {
  pname = "tmux-rime";
  version = "0.0.1";
  src = fetchFromGitHub {
    owner = "Freed-Wu";
    repo = pname;
    rev = version;
    fetchSubmodules = false;
    sha256 = "sha256-zZ6+sZhJW91/TiF6aR0n44jCj2iEkL+RzdEVTw6G9K8=";
  };

  nativeBuildInputs = [ glib.dev stdenv.cc unzip pkg-config xmake ];
  buildInputs = [ glib librime ];

  buildPhase = ''
    XMAKE_ROOT=y xmake
  '';

  installPhase = ''
    XMAKE_ROOT=y xmake install -o$out
  '';

  meta = with lib; {
    homepage = "https://github.com/Freed-Wu/tmux-rime";
    description = "rime for tmux";
    license = licenses.gpl3;
    maintainers = with maintainers; [ Freed-Wu ];
    platforms = platforms.unix;
  };
}
```

**Reply by waruqi** [reply]
I have not nix os.

**Comment by eihqnh** [comment, marked as answer]
I m not sure if this will work. 
https://github.com/wineee/nur-packages/blob/main/pkgs/mogan/default.nix
Thanks to wineee
(Maybe we could genernate Cmakelist?)

**Reply by Freed-Wu** [reply]
https://github.com/wineee/nur-packages/blob/main/pkgs/mogan/default.nix#L71 :smile:

I think xmake needs git is weird. even we use private network. Does it a bug?

```
       > note: install or modify (m) these packages (pass -y to skip confirm)?
       > in builtin-repo:
       >   -> git latest
       > please input: y (y/n/m)
```

**Reply by eihqnh** [reply]
> https://github.com/wineee/nur-packages/blob/main/pkgs/mogan/default.nix#L71 :smile:
> 
> I think xmake needs git is weird. even we use private network. Does it a bug?
> 
> ```
>        > note: install or modify (m) these packages (pass -y to skip confirm)?
>        > in builtin-repo:
>        >   -> git latest
>        > please input: y (y/n/m)
> ```

Actually,I don't know how it works inside....
Maybe we need to add xmake build support to nixpkgs?

**Reply by Freed-Wu** [reply]
`xmake g --network=private` must be set to avoid network bug.
and I meet

```
       > note: the following packages were not found in any repository (check if they are spelled correctly):
       >   -> glib
```

Even if both pkg-config and glib are available.

--verbose said:

```
       > /nix/store/pbixrqd120nnw8a0ygfpm3mnwlr1mbj7-binutils-2.43.1/bin/ld: cannot find -lglib: No such file or directory
       > collect2: error: ld returned 1 exit status
```

**Reply by Freed-Wu** [reply]
OK, I can conclude:

1. xmake need git, even if you don't need any extern source.

> note: install or modify (m) these packages (pass -y to skip confirm)?
> in builtin-repo:
>   -> git latest
> please input: y (y/n/m)

2. `xmake g --network=private` can avoid download artifact, `xmake f --policies=package.precompiled:n` still download artifact (perhaps this is why we need git?)

> Cloning into '/build/tmp.mIE3fLfzGE/.xmake/repositories/build-artifacts'...
> fatal: unable to access 'https://github.com/xmake-mirror/build-artifacts.git/': Could not resolve host: github.com
> error: execv(git clone https://github.com/xmake-mirror/build-artifacts.git -b main -c core.fsmonitor=false -c core.autocrlf=false /build/tmp.mIE3fLfzGE/.xmake/repositories/build-artifacts) failed(128)

You must set `xmake g --network=private` at first step.

3. pacakge name failed. `add_package('glib')`'s `glib` means `pkg-config glib-2.0`. However, after `xmake g --network=private`, it search `pkg-config glib`.
In other Linux distribution, the behaviour is same.

So we should:

1. add git to nativeBuildInputs
2. set private network at first step
3. `substituteInPlace xmake.lua --replace "glib" "glib-2.0"`

**Reply by Freed-Wu** [reply]
I think 1 and 3 are incorrect behavior.

**Reply by Freed-Wu** [reply]
1. git should be an optional package.
2. The best steps is we let nixpkgs support xmake, we just add xmake to nativeBuildInputs, it will run `xmake g --network=private`, `xmake`, `xmake install` in correct phase, like add `cmake` will run `cmake -B`, `cmake --build`, `cmake --install`. It can simplify build script for 2.
3. xmake support nix, it will know glib means nixpkgs's `pkgs.glib`, and search the dir of `pkgs.glib` to find a `glib-2.0.pc`. It can solve 3.

**Reply by Freed-Wu** [reply]
Not nix has the problem 2. other distribution also lack support of xmake. Such as in gentoo and yocto, we can `inherit cmake`. However. We cannot `inherit xmake`.

**Reply by Freed-Wu** [reply]
Better for debug, `xmake f --verbose` can be add after `xmake g --network=private`.

**Reply by Freed-Wu** [reply]
Also `export XMAKE_ROOT=y HOME=$(mktemp -d)`. package manager ususally run in root, and xmake will use `$HOME/.xmake` to restore config for `xmake g`.

```
error: save /homeless-shelter/.xmake/xmake.conf failed, cannot open file: /homeless-shelter/.xmake/xmake.conf, No such file or directory!
```

**Comment by Freed-Wu** [comment]
git is unnecessary. Look at this black magic:

```
    export XMAKE_ROOT=y
    HOME=$PWD PATH=$HOME:$PATH
    echo -e "#!$SHELL\necho I am git" > $HOME/git
    chmod +x $HOME/git
    xmake g --network=private
    xmake f --verbose
```

**Reply by Freed-Wu** [reply]
xmake only check `git --version`.

**Comment by star-hengxing** [comment]
> substituteInPlace xmake.lua --replace "glib" "glib-2.0"

We need to add `add_extsources("pkgconfig::glib-2.0")` in xrepo glib package.
https://github.com/xmake-io/xmake-repo/pull/5672

**Reply by Freed-Wu** [reply]
If we don't use private network, xmake can know glib is glib-2.0.

**Reply by waruqi** [reply]
if `add_requires` is used in user project, git will be used. But if git has been installed on system/$PATH, xmake will find and use it first.

**Comment by Freed-Wu** [comment]
For linux distribution packagers: there are some examples to package a project which uses xmake as its build system to many common linux distributions:

- [ArchLinux](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=tmux-rime)
- [Gentoo Linux](https://gitweb.gentoo.org/repo/proj/guru.git/tree/app-misc/tmux-rime/tmux-rime-0.0.3.ebuild?h=dev)
- [NixOS](https://github.com/Freed-Wu/nur-packages/blob/main/pkgs/tools/misc/tmux-rime/default.nix)
- [Ubuntu](https://github.com/Freed-Wu/tmux-rime/blob/main/debian/rules)
